### PR TITLE
5 only one level of sub machine supported

### PR DESCRIPTION
**Pull request by 9Fingered**
Daniel -  This is a massive PR.  Breaking it up would be confusing, though, as some of the earlier commits are undone by later commits.  Here are the major topics:

- in the fsm build itself, use -MG when creating dependency files to add missing headers to the dependency list, rather than complain about them.
-  modify printAncestry to print multiple generations
  - selectors are added to chose upper or lower case and to tweak how many generations are printed.
- make output generator factories which know how to create a generator for the different output file types.  The factory can create a "top level" generator or a "sub machine" generator, as requested.  This is the key to processing more than one sub-machine level, as it removes the use of static data for the sub machine generators.
  - Along these lines, expand the use of the pCMachineData struct (as well as the struct itself)
  - Create "C Language Parts" functions to build (and store) the strings for the different language constructs.
    - These are adjusted for the case where only one level of sub-machine is present (to maintain compatibility with current code generation) when new cl option, --short-user-fn-names, is used.
  - Use the "THIS" macro to shorten code lines.
  - Create and use new "UFMN" (User Function Machine Name) macro which is sensitive to --short-user-fn-names.
- add an option family (-M, -Mh, -Md) so that fsmlang will play nicely with user make files.
  - -M and -Mh simply give a list of the source or header files, respectively, that would be produced with the accompanying -t option (which must be given).  This is most useful in target dependency rules or the clean target.
  - -Md will give a make file suitable multi-target recipe showing how to create the output files from the .fsm on which they depend.
  - The use of these options has been incorporated into the make files for many of the full tests.
- html and plantuml print ancestry in certain cases where it helps to show proper relationships.  Where ancestry is printed, the :: separator is used.
- The C (both flavors) header generation now creates multiple headers when sub-machines are involved.  The header bearing the name of the input file is the "public" header to be included by files *using* the machine.  Other headers are for "internal" use only.  The idea is that some definitions can be made shorter by limiting their scope.  More might be done here.
- Related to the above, some material previously included in the header files is now put into the source file. Some functions are now declared to be static.
- No RUN... macro is provided; the machine invoking function is always created.
- New cl option, --short-user-fn-names[=<*true|false>] is provided (as mentioned above).  When used, the name prefix for user functions (actions, transition functions, _etc._) is the plain machine name.

### CODE REVIEW

**Review by Jasinsky — APPROVED** [review]
Looks good, for some reason I am missing my previous review comments - but there were very minor things.

File: src/ancestry.c [hunk 1]
```diff
@@ -0,0 +1,978 @@
...
+*    File created by Steven Stanton
+*
+*  Long Description:
+*
+*/
+
+#include "ancestry.h"
+
+#if defined (CYGWIN) || defined (LINUX)
+	#include <stdio.h>
```
[the remark this answers is not in the document]
Comment (9Fingered) [reply src/ancestry.c:39]: At one time .... :{)

File: src/ancestry.c [hunk 2]
```diff
@@ -0,0 +1,978 @@
...
+	#include <time.h>
+#endif
+#include <string.h>
+#include <stdlib.h>
+
+char *createAncestryFileName(pMACHINE_INFO pmi)
+{
+	FILE *tmp = tmpfile();
+	char *cp  = NULL;
+	unsigned long file_size;
```
[the remark this answers is not in the document]
Comment (9Fingered) [reply src/ancestry.c:53]: What if tmp turns out to be NULL?

[the remark this answers is not in the document]
Comment (9Fingered) [reply src/ancestry.c:53]: Got it; -1.  all good.

File: src/ancestry.c [hunk 3]
```diff
@@ -0,0 +1,978 @@
...
+ *  Behaves similarly to hungarianToUnderbarCaps, but writes
+ *  the ouput to the given stream.
+*/
+void streamHungarianToUnderbarCaps(FILE *fout, char *str)
+{
+
+	int 	consecutive;
+	char	*cp;
+
+	consecutive = 0;
```
[the remark this answers is not in the document]
Comment (9Fingered) [reply src/ancestry.c:160]: The construction was from a very old style preference at a former company.  I like your suggested construction better, myself.

File: src/ancestry.c [hunk 4]
```diff
@@ -0,0 +1,978 @@
...
+
+*/
+char *hungarianToUnderbarCaps(char *str)
+{
+
+	int 	i,consecutive;
+	char	*cp, *cp1, *cp2;
+
+	/* first, use i and cp1 to figure out how much memory to get */
+	i = strlen(str);
```
[the remark this answers is not in the document]
Comment (9Fingered) [reply src/ancestry.c:96]: Yes, why not?

File: src/ancestry.c [hunk 5]
```diff
@@ -0,0 +1,978 @@
...
+			fputc('_',fout);
+			continue;
+		}
+
+		if (
+          !(*cp & 0x20)
+          && (*cp != '_')
+		  && (cp != str)
+          )
+        {
```
[the remark this answers is not in the document]
Comment (9Fingered) [reply src/ancestry.c:175]: Beautified; hopefully resolves.

File: src/ancestry.c
[the remark this answers is not in the document]
Comment (9Fingered) [reply src/ancestry.c:1]: The repetition comes from copy/pasting the original function.  But, your suggestion is good.

Are there other things we would add to such a module?

File: src/parser.y [hunk 6]
```diff
@@ -46,9 +46,10 @@ extern int yylex(void);
 
 char *rindex(const char *str,int c);
 
-pMACHINE_INFO          pmachineInfo = NULL;
-pFSMOutputGenerator	  pfsmog;
-pSTATE_AND_EVENT_DECLS psedecls = NULL;
+pMACHINE_INFO               pmachineInfo = NULL;
+pFSMOutputGenerator	        pfsmog       = NULL;
```
[the remark this answers is not in the document]
Comment (9Fingered) [reply src/parser.y:50]: Inconsistent use of tabs/spaces.  In the old, old days, we had not realized how bad tabs were :{)

File: src/fsm_priv.h [hunk 7]
```diff
@@ -38,7 +38,13 @@
 
 #include "list.h"
 
-/*
+#ifdef FSMLANG_DEVELOP
```
[the remark this answers is not in the document]
Comment (9Fingered) [reply src/fsm_priv.h:41]: Good idea.  Let's design such.

File: src/ancestry.c
[the remark this answers is not in the document]
Comment (9Fingered) [reply src/ancestry.c:1]: Not being able to think of other items to add to a string utilities module, I added a static function to ancestry.c to do the string extraction from the file.

File: src/fsm_priv.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (9Fingered) [reply src/fsm_priv.h:41]: Let's resolve this conversation by noting that you've added an issue to track the creation of such a logger subsystem.

**Comment by 9Fingered** [comment]
Ready for final review/approval.

**Review by Jasinsky — APPROVED** [review]

**Pull request merged**